metric/metric.go

```diff
@@ -2,26 +2,161 @@ package metric
+			Help:      "Number of results per endpoint",
+		}, []string{"key", "group", "name", "url", "success"})
+
+		resultSuccessGauge = promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "results_success",
+			Help:      "Displays whether or not the watchdog result was a success",
+		}, []string{"key", "group", "name", "url"})
+
+		resultDurationGauge = promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "results_duration_seconds",
+			Help:      "Returns how long the watchdog took to complete in seconds",
+		}, []string{"key", "group", "name", "url"})
+
+		resultDNSReturnCode = promauto.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "results_dns_return_code",
+			Help:      "Response DNS return code",
+		}, []string{"key", "group", "name", "url", "code"})
+
+		resultTCPConnected = promauto.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "results_tcp_connected",
+			Help:      "Response TCP connected count",
+		}, []string{"key", "group", "name", "url"})
+
+		resultStartTLSConnected = promauto.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "results_starttls_connected",
+			Help:      "Response StartTLS connected count",
+		}, []string{"key", "group", "name", "url"})
+
+		resultStartTLSSSLLastChainExpiryTimestampSeconds = promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "results_starttls_ssl_last_chain_expiry_timestamp_seconds",
+			Help:      "Returns last SSL chain expiry in timestamp seconds",
+		}, []string{"key", "group", "name", "url"})
+
+		resultTLSConnected = promauto.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "results_tls_connected",
+			Help:      "Response TLS connected count",
+		}, []string{"key", "group", "name", "url"})
+
+		resultTLSSSLLastChainExpiryTimestampSeconds = promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "results_tls_ssl_last_chain_expiry_timestamp_seconds",
+			Help:      "Returns last SSL chain expiry in timestamp seconds",
+		}, []string{"key", "group", "name", "url"})
+
+		resultHTTPConnected = promauto.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "results_http_connected",
+			Help:      "Response HTTP connected count",
+		}, []string{"key", "group", "name", "url"})
+
+		resultHTTPStatusCode = promauto.NewCounterVec(prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "results_http_status_code",
+			Help:      "Response HTTP status code",
+		}, []string{"key", "group", "name", "url", "code"})
+
+		resultHTTPSSLLastChainExpiryTimestampSeconds = promauto.NewGaugeVec(prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "results_http_ssl_last_chain_expiry_timestamp_seconds",
+			Help:      "Returns last SSL chain expiry in timestamp seconds",
+		}, []string{"key", "group", "name", "url"})
```

Comment:
I feel like a lot of these metrics are unnecessary.

For instance, why do we need `results_success` if `results_total` already has a label `success`?

`results_tcp_connected`, `results_starttls_connected`, `results_tls_connected`, `results_http_connected` and `results_http_status_code` should all be suffixed with `_total` for consistency.

But before even looking too deeply into that, does it really make sense to have all of these individual counters?

You can technically determine the type by simply looking at the url, granted your query might look weird, but even without resorting to this hack, we could just add the label `type` to `results_total` where the type is `http`, `starttls`, `dns`, `tcp` or `icmp`, and we would no longer need most of the metrics you added.

Reply:
I think you're right. After added the type label in each metrics. Many metrics can be merged.